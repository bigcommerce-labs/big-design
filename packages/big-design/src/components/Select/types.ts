import { Placement } from 'popper.js';

import { ListItemProps } from '../List/Item';

export interface SelectProps extends Omit<React.HTMLAttributes<HTMLUListElement>, 'children' | 'onChange'> {
  action?: Action;
  disabled?: boolean;
  error?: string;
  label?: string;
  maxHeight?: number;
  multi?: boolean;
  options: Option[];
  placement?: Placement;
  positionFixed?: boolean;
  required?: boolean;
  value?: string | number | Array<string | number>;
  onChange(value: string | number | Array<string | number>): void;
}

interface BaseItem extends Omit<ListItemProps, 'children' | 'content' | 'onClick' | 'value'> {
  content: string;
  icon?: React.ReactElement;
}

export interface Option extends Omit<BaseItem, 'actionType'> {
  value: string | number;
}

export interface Action extends BaseItem {
  onClick(inputText: string): void;
}
