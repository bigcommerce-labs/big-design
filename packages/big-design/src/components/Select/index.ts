export { Select } from './Select';
export {
  type SelectAction,
  type SelectOption,
  type SelectOptionGroup,
  type SelectProps,
} from './types';
