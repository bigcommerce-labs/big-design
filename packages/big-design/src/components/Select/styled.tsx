import { hideVisually } from 'polished';
import styled from 'styled-components';

export const StyledStatusMessage = styled.div`
  ${hideVisually()}
`;
