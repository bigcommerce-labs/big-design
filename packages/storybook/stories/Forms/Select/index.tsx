export * from './Select';
export * from './SelectPropTable';
