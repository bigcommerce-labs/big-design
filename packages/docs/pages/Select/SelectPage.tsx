import { Form, FormGroup, Grid, H0, H1, H2, Link, Select, Text } from '@bigcommerce/big-design';
import { DeleteIcon } from '@bigcommerce/big-design-icons';
import React, { useState } from 'react';

import { Code, CodePreview } from '../../components';
import { SelectActionPropTable, SelectGroupPropTable, SelectOptionPropTable, SelectPropTable } from '../../PropTables';

const SelectPage = () => (
  <>
    <H0>Selects</H0>

    <Text>
      Select are typeable inputs with selectable dropdown items.{' '}
      <Link href="https://design.bigcommerce.com/components/selects" target="_blank">
        Selects Design Guidelines
      </Link>
      .
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      {function Example() {
        const [value, setValue] = useState('mx');
        const handleChange = (val) => setValue(val);

        return (
          <Form>
            <FormGroup>
              <Select
                action={{
                  actionType: 'destructive' as 'destructive',
                  content: 'Remove Country',
                  icon: <DeleteIcon />,
                  onActionClick: () => null,
                }}
                filterable={true}
                label="Countries"
                maxHeight={300}
                onOptionChange={handleChange}
                options={[
                  { value: 'us', content: 'United States' },
                  { value: 'mx', content: 'Mexico' },
                  { value: 'ca', content: 'Canada' },
                  { value: 'en', content: 'England' },
                  { value: 'fr', content: 'France' },
                  { value: 'gr', content: 'Germany' },
                  { value: 'ar', content: 'Argentina' },
                  { value: 'ru', content: 'Russia', disabled: true },
                  { value: 'ch', content: 'Chile' },
                  { value: 'bo', content: 'Bolivia' },
                  { value: 'jp', content: 'Japan' },
                  { value: 'cn', content: 'China' },
                  { value: 'sk', content: 'South Korea' },
                  { value: 'au', content: 'Australia' },
                  { value: 'ug', content: 'Uganda' },
                ]}
                placeholder={'Choose country'}
                placement={'bottom-start'}
                required
                value={value}
              />
            </FormGroup>
          </Form>
        );
      }}
      {/* jsx-to-string:end */}
    </CodePreview>

    <H1>API</H1>
    <SelectPropTable />
    <SelectOptionPropTable />
    <SelectActionPropTable />
    <SelectGroupPropTable />

    <H1>Examples</H1>

    <H2>Position</H2>

    <Text>
      Select can be anchored in different directions with the <Code primary>position</Code> property. It will
      automatically find a position if there's not enough space in the chosen direction.
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      <Grid gridColumns="repeat(4, 1fr)">
        <Select
          label="Select"
          onOptionChange={() => null}
          options={[
            { value: 1, content: 'Option' },
            { value: 2, content: 'Option' },
            { value: 3, content: 'Option' },
            { value: 4, content: 'Option' },
          ]}
          placeholder="Choose from above"
          placement="top"
          required
        />
        <Select
          label="Select"
          onOptionChange={() => null}
          options={[
            { value: 1, content: 'Option' },
            { value: 2, content: 'Option' },
            { value: 3, content: 'Option' },
            { value: 4, content: 'Option' },
          ]}
          placeholder="Choose from below"
          placement="bottom-start"
          required
        />
        <Select
          label="Select"
          onOptionChange={() => null}
          options={[
            { value: 1, content: 'Option' },
            { value: 2, content: 'Option' },
            { value: 3, content: 'Option' },
            { value: 4, content: 'Option' },
          ]}
          placeholder="Choose from the right"
          placement="right-start"
          required
        />
        <Select
          label="Select"
          onOptionChange={() => null}
          options={[
            { value: 1, content: 'Option' },
            { value: 2, content: 'Option' },
            { value: 3, content: 'Option' },
            { value: 4, content: 'Option' },
          ]}
          placeholder="Choose from the left"
          placement="left-end"
          required
        />
      </Grid>
      {/* jsx-to-string:end */}
    </CodePreview>

    <H2>Max Height</H2>

    <Text>
      Once the content is longer than the max-height, the Dropdown will be scrollable. It is possible to modify the
      dimension by passing a <Code primary>maxHeight</Code> property.
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      <Grid gridColumns="repeat(3, 1fr)">
        <Select
          label="Select"
          onOptionChange={() => null}
          options={[
            { value: 1, content: 'Option' },
            { value: 2, content: 'Option' },
            { value: 3, content: 'Option' },
            { value: 4, content: 'Option' },
            { value: 5, content: 'Option' },
            { value: 6, content: 'Option' },
            { value: 7, content: 'Option' },
            { value: 8, content: 'Option' },
          ]}
          placeholder="Default"
          required
        />
        <Select
          label="Select"
          maxHeight={150}
          onOptionChange={() => null}
          options={[
            { value: 1, content: 'Option' },
            { value: 2, content: 'Option' },
            { value: 3, content: 'Option' },
            { value: 4, content: 'Option' },
            { value: 5, content: 'Option' },
            { value: 6, content: 'Option' },
            { value: 7, content: 'Option' },
            { value: 8, content: 'Option' },
          ]}
          placeholder="Smaller"
          required
        />
        <Select
          label="Select"
          maxHeight={350}
          onOptionChange={() => null}
          options={[
            { value: 1, content: 'Option' },
            { value: 2, content: 'Option' },
            { value: 3, content: 'Option' },
            { value: 4, content: 'Option' },
            { value: 5, content: 'Option' },
            { value: 6, content: 'Option' },
            { value: 7, content: 'Option' },
            { value: 8, content: 'Option' },
            { value: 9, content: 'Option' },
            { value: 10, content: 'Option' },
            { value: 11, content: 'Option' },
            { value: 12, content: 'Option' },
          ]}
          placeholder="Larger"
          required
        />
      </Grid>
      {/* jsx-to-string:end */}
    </CodePreview>

    <H2>Disabled</H2>

    <Text>
      It is possible to disable the entire select component, similar to how you would disable a native HTML select
      element, by using the <Code primary>disabled</Code> property.
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      <Form>
        <FormGroup>
          <Select
            disabled
            label="Select"
            maxHeight={350}
            onOptionChange={() => null}
            options={[
              { value: 1, content: 'Option' },
              { value: 2, content: 'Option' },
              { value: 3, content: 'Option' },
              { value: 4, content: 'Option' },
            ]}
            placeholder="Larger"
            required
          />
        </FormGroup>
      </Form>

      {/* jsx-to-string:end */}
    </CodePreview>

    <H2>Action</H2>

    <Text>
      Accepts an <Code>action</Code> object to display at the end of the list.
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      <Form>
        <FormGroup>
          <Select
            action={{
              actionType: 'destructive',
              content: 'Remove Country',
              icon: <DeleteIcon />,
              onActionClick: () => null,
            }}
            label="Countries"
            onOptionChange={() => null}
            options={[
              { value: 'us', content: 'United States' },
              { value: 'mx', content: 'Mexico' },
              { value: 'ca', content: 'Canada' },
              { value: 'en', content: 'England' },
            ]}
            placeholder={'Choose country'}
            placement={'bottom-start'}
            required
          />
        </FormGroup>
      </Form>
      {/* jsx-to-string:end */}
    </CodePreview>

    <H2>Error</H2>

    <Text>
      An <Code>error</Code> prop receives a <Code>string</Code> to display.
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      <Form>
        <FormGroup>
          <Select
            label="Countries"
            error="Need to choose a country before proceeding"
            onOptionChange={() => null}
            options={[
              { value: 'us', content: 'United States' },
              { value: 'mx', content: 'Mexico' },
              { value: 'ca', content: 'Canada' },
              { value: 'en', content: 'England' },
            ]}
            placeholder={'Choose country'}
            placement={'bottom-start'}
            required
          />
        </FormGroup>
      </Form>
      {/* jsx-to-string:end */}
    </CodePreview>

    <H2>Select Groups</H2>

    <Text>
      It is possible to create grouped options with labels with a <Code primary>Select Group</Code> by passing{' '}
      <Code primary>label</Code> and <Code primary>options</Code> to the top-level <Code primary>options</Code>{' '}
      property.
    </Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      <Form>
        <FormGroup>
          <Select
            label="My Options"
            filterable
            value="ca"
            onOptionChange={() => null}
            options={[
              {
                label: 'Group 1',
                options: [
                  { value: 'mx', content: 'Mexico' },
                  { value: 'ca', content: 'Canada' },
                  { value: 'en', content: 'England' },
                ],
              },
              {
                label: 'Group 2',
                options: [
                  { value: 'fr', content: 'France' },
                  { value: 'gr', content: 'Germany' },
                  { value: 'ar', content: 'Argentina' },
                ],
              },
            ]}
          />
        </FormGroup>
      </Form>
      {/* jsx-to-string:end */}
    </CodePreview>

    <H2>Option & Action Description</H2>

    <Text>It is possible to add a description for select options and actions.</Text>

    <CodePreview>
      {/* jsx-to-string:start */}
      {function Example() {
        const [value, setValue] = useState(1);
        const handleChange = (val) => setValue(val);

        return (
          <Form>
            <FormGroup>
              <Select
                action={{
                  actionType: 'destructive',
                  content: 'Remove',
                  description: 'Description for remove action',
                  icon: <DeleteIcon />,
                  onActionClick: () => null,
                }}
                label="Select"
                onOptionChange={() => null}
                options={[
                  { value: 1, content: 'Option #1', description: 'Description for option #1' },
                  {
                    value: 2,
                    content: 'Option #2',
                    description: 'Description for option #2',
                    disabled: true,
                  },
                  { value: 3, content: 'Option #3' },
                  { value: 4, content: 'Option #4' },
                  { value: 5, content: 'Option #5' },
                ]}
                onChange={handleChange}
                placeholder="Larger"
                required
                value={value}
              />
            </FormGroup>
          </Form>
        );
      }}
      {/* jsx-to-string:end */}
    </CodePreview>
  </>
);

export default SelectPage;
