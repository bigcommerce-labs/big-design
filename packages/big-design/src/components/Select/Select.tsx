import { useCombobox, UseComboboxState, UseComboboxStateChangeOptions } from 'downshift';
import React, {
  cloneElement,
  createRef,
  isValidElement,
  RefObject,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { usePopper } from 'react-popper';

import { useUniqueId } from '../../hooks';
import { typedMemo, warning } from '../../utils';
import { Box } from '../Box';
import { FormControlLabel } from '../Form';
import { Input } from '../Input';
import { List } from '../List';
import { SelectOption, SelectOptionGroup, SelectProps } from '../Select';
import { DropdownButton, StyledDropdownIcon, StyledInputContainer } from '../Select/styled';
import { SelectAction } from '../Select/types';

export const Select = typedMemo(
  <T extends unknown>({
    action,
    autoComplete = 'off',
    autoWidth = false,
    className,
    disabled,
    filterable = true,
    id,
    inputRef,
    label,
    labelId,
    maxHeight,
    onOptionChange,
    options,
    placeholder,
    placement = 'bottom-start' as const,
    positionFixed = false,
    required,
    style,
    value,
    ...props
  }: SelectProps<T>): ReturnType<React.FC<SelectProps<T>>> => {
    const defaultRef: RefObject<HTMLInputElement> = createRef();
    const selectUniqueId = useUniqueId('select');

    const [inputValue, setInputValue] = useState<string | undefined>('');

    const flattenOptions = useCallback((options: SelectProps<T>['options']) => {
      const isGroups = (
        options: Array<SelectOptionGroup<T> | SelectOption<T>>,
      ): options is Array<SelectOptionGroup<T>> =>
        options.every((option) => 'options' in option && !('value' in option));

      return isGroups(options)
        ? options.map((group) => group.options).reduce((acum, curr) => acum.concat(curr), [])
        : options;
    }, []);

    // We need to pass Downshift only options without groups for accessibility tracking
    const flattenedOptions = useMemo(() => (action ? [...flattenOptions(options), action] : flattenOptions(options)), [
      action,
      flattenOptions,
      options,
    ]);

    // Find the selected option
    const selectedOption = useMemo(() => {
      return flattenedOptions.find((option): option is SelectOption<T> => 'value' in option && option.value === value);
    }, [flattenedOptions, value]);

    // Initialize with flattened options
    const [filteredOptions, setFilteredOptions] = useState(flattenedOptions);

    // Need to set select options if options prop changes
    useEffect(() => setFilteredOptions(flattenedOptions), [flattenedOptions]);

    const handleOnSelectedItemChange = (changes: Partial<UseComboboxState<SelectOption<T> | SelectAction | null>>) => {
      if (action && changes.selectedItem && changes.selectedItem.content === action.content) {
        action.onActionClick(inputValue || null);
      } else if (changes.selectedItem && 'value' in changes.selectedItem && typeof onOptionChange === 'function') {
        onOptionChange(changes.selectedItem.value, changes.selectedItem);
      }
    };

    const handleOnInputValueChange = ({
      inputValue,
      isOpen,
    }: Partial<UseComboboxState<SelectOption<T> | SelectAction | null>>) => {
      // Filter only when List is open
      if (filterable && isOpen === true) {
        setFilteredOptions(filterOptions(inputValue));
      }

      setInputValue(inputValue || '');
    };

    const filterOptions = (inputVal = '') => {
      return flattenedOptions.filter(
        (option) =>
          option.content === (action && action.content) ||
          option.content.toLowerCase().startsWith(inputVal.trim().toLowerCase()),
      );
    };

    const handleOnIsOpenChange = ({ isOpen }: Partial<UseComboboxState<SelectOption<T> | SelectAction | null>>) => {
      if (filterable && isOpen === false) {
        // Reset the options when the List is closed
        setFilteredOptions(flattenedOptions);
      }
    };

    const handleStateReducer = (
      _state: UseComboboxState<SelectOption<T> | SelectAction | null>,
      actionAndChanges: UseComboboxStateChangeOptions<SelectOption<T> | SelectAction | null>,
    ) => {
      switch (actionAndChanges.type) {
        case useCombobox.stateChangeTypes.InputBlur:
          return { ...actionAndChanges.changes, inputValue: selectedOption ? selectedOption.content : '' };
        default:
          return actionAndChanges.changes;
      }
    };

    const {
      closeMenu,
      getComboboxProps,
      getInputProps,
      getItemProps,
      getLabelProps,
      getMenuProps,
      getToggleButtonProps,
      highlightedIndex,
      isOpen,
      openMenu,
      selectedItem,
    } = useCombobox({
      id: selectUniqueId,
      inputId: id,
      inputValue,
      itemToString: (item) => (item ? item.content : ''),
      items: filteredOptions,
      labelId,
      onInputValueChange: handleOnInputValueChange,
      onIsOpenChange: handleOnIsOpenChange,
      onSelectedItemChange: handleOnSelectedItemChange,
      selectedItem: selectedOption || null,
      stateReducer: handleStateReducer,
    });

    // Popper
    const referenceRef = useRef(null);
    const popperRef = useRef(null);

    const { styles, attributes, update } = usePopper(referenceRef.current, popperRef.current, {
      modifiers: [
        {
          name: 'eventListeners',
          options: {
            scroll: isOpen,
            resize: isOpen,
          },
        },
        {
          name: 'offset',
          options: {
            offset: [0, 4],
          },
        },
      ],
      strategy: positionFixed ? 'fixed' : 'absolute',
      placement,
    });

    const setCallbackRef = useCallback(
      (ref: HTMLInputElement) => {
        if (typeof inputRef === 'function') {
          inputRef(ref);
        }
      },
      [inputRef],
    );

    const getInputRef = useCallback(() => {
      if (inputRef && typeof inputRef === 'object') {
        return inputRef;
      } else if (typeof inputRef === 'function') {
        return setCallbackRef;
      }

      return defaultRef;
    }, [defaultRef, inputRef, setCallbackRef]);

    const renderLabel = useMemo(() => {
      if (!label) {
        return null;
      }

      if (typeof label === 'string') {
        return (
          <FormControlLabel {...getLabelProps()} renderOptional={!required}>
            {label}
          </FormControlLabel>
        );
      }

      if (isValidElement(label) && label.type === FormControlLabel) {
        return cloneElement(label as React.ReactElement<React.LabelHTMLAttributes<HTMLLabelElement>>, getLabelProps());
      }

      warning('label must be either a string or a FormControlLabel component.');
    }, [getLabelProps, label, required]);

    const renderToggle = useMemo(() => {
      return (
        <DropdownButton
          {...getToggleButtonProps({
            disabled,
            type: 'button',
          })}
          aria-label="toggle menu"
          variant="subtle"
        >
          <StyledDropdownIcon />
        </DropdownButton>
      );
    }, [disabled, getToggleButtonProps]);

    const renderInput = useMemo(() => {
      return (
        <StyledInputContainer ref={referenceRef}>
          <Input
            {...getInputProps({
              ...props,
              autoComplete,
              disabled,
              onClick: () => {
                !isOpen && openMenu();
              },
              onFocus: () => {
                !isOpen && openMenu();
              },
              onKeyDown: (event) => {
                switch (event.key) {
                  case 'Enter':
                    event.preventDefault();
                    if (isOpen === false) {
                      openMenu();
                      // https://github.com/downshift-js/downshift/issues/734
                      (event.nativeEvent as any).preventDownshiftDefault = true;
                    }
                    break;
                  case 'Escape':
                    if (isOpen === false) {
                      // Reset the value to empty
                      onOptionChange();
                    } else {
                      closeMenu();
                    }
                    // https://github.com/downshift-js/downshift/issues/734
                    (event.nativeEvent as any).preventDownshiftDefault = true;
                    break;
                }
              },
              placeholder,
              ref: getInputRef(),
              readOnly: !filterable,
              required: required,
            })}
            iconLeft={selectedItem?.icon}
            iconRight={renderToggle}
          />
        </StyledInputContainer>
      );
    }, [
      autoComplete,
      closeMenu,
      disabled,
      filterable,
      getInputProps,
      getInputRef,
      isOpen,
      onOptionChange,
      openMenu,
      placeholder,
      props,
      renderToggle,
      required,
      selectedItem,
    ]);

    return (
      <div>
        {renderLabel}
        <div {...getComboboxProps()}>{renderInput}</div>
        <Box ref={popperRef} style={styles.popper} {...attributes.poppper} zIndex="popover">
          <List
            action={action}
            autoWidth={autoWidth}
            filteredItems={filteredOptions}
            getItemProps={getItemProps}
            getMenuProps={getMenuProps}
            highlightedIndex={highlightedIndex}
            isOpen={isOpen}
            items={options}
            maxHeight={maxHeight}
            selectedItem={selectedItem && 'value' in selectedItem ? selectedItem : null}
            update={update}
          />
        </Box>
      </div>
    );
  },
);
