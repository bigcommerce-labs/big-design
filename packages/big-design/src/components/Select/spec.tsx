import 'jest-styled-components';
import React from 'react';
import { fireEvent, render } from 'react-testing-library';

import { Select } from './Select';

jest.mock('popper.js', () => {
  const PopperJS = jest.requireActual('popper.js');

  return class {
    static placements = PopperJS.placements;

    constructor() {
      return {
        // tslint:disable-next-line: no-empty
        destroy: () => {},
        // tslint:disable-next-line: no-empty
        scheduleUpdate: () => {},
      };
    }
  };
});

const onItemChange = jest.fn();

const SelectMock = (
  <Select onItemChange={onItemChange} label="Countries" placeholder="Choose country">
    <Select.Option value="us">United States</Select.Option>
    <Select.Option value="mx">Mexico</Select.Option>
    <Select.Option value="ca">Canada</Select.Option>
    <Select.Option value="en">England</Select.Option>
  </Select>
);

test('renders select combobox', () => {
  const { getByRole } = render(SelectMock);

  expect(getByRole('combobox')).toBeInTheDocument();
});

test('renders select label', () => {
  const { getByText } = render(SelectMock);

  expect(getByText('Countries')).toBeInTheDocument();
});

test('select label has id', () => {
  const { getByText } = render(SelectMock);

  expect(getByText('Countries').id).toBeDefined();
});

test('select label has for attribute', () => {
  const { getByText } = render(SelectMock);

  expect(getByText('Countries').hasAttribute('for')).toBe(true);
});

test('renders select input', () => {
  const { getByLabelText } = render(SelectMock);

  expect(getByLabelText('Countries')).toBeInTheDocument();
});

test('select input has id', () => {
  const { getByLabelText } = render(SelectMock);

  expect(getByLabelText('Countries').id).toBeDefined();
});

test('select input has placeholder text', () => {
  const { getByPlaceholderText } = render(SelectMock);

  expect(getByPlaceholderText('Choose country')).toBeDefined();
});

test('select input has aria-autocomplete', () => {
  const { getByLabelText } = render(SelectMock);

  expect(getByLabelText('Countries').getAttribute('aria-autocomplete')).toBe('list');
});

test('select input has aria-labelledby', () => {
  const { getByLabelText, getByText } = render(SelectMock);
  const input = getByLabelText('Countries');
  const label = getByText('Countries');

  expect(input.getAttribute('aria-labelledby')).toBe(label.id);
});

test('select input has aria-controls', () => {
  const { getByRole, getByLabelText } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);

  expect(input.getAttribute('aria-controls')).toBe(getByRole('listbox').id);
});

test('renders input button', () => {
  const { getByRole } = render(SelectMock);

  expect(getByRole('button'));
});

test('input button has aria-haspopup', () => {
  const { getByRole } = render(SelectMock);

  expect(getByRole('button').getAttribute('aria-haspopup')).toBe('true');
});

test('input button has aria-label', () => {
  const { getByRole } = render(SelectMock);

  expect(getByRole('button').getAttribute('aria-label')).toBe('toggle menu');
});

test('select menu opens when focused on input', () => {
  const { getByLabelText, queryByRole } = render(SelectMock);
  const input = getByLabelText('Countries');

  expect(queryByRole('listbox')).not.toBeInTheDocument();

  fireEvent.focus(input);

  expect(queryByRole('listbox')).toBeInTheDocument();
});

test('select menu opens/closes when input button is clicked', () => {
  const { getByRole, queryByRole } = render(SelectMock);
  const button = getByRole('button');

  fireEvent.click(button);
  expect(queryByRole('listbox')).toBeInTheDocument();

  fireEvent.click(button);
  expect(queryByRole('listbox')).not.toBeInTheDocument();
});

test('select has items', () => {
  const { getAllByRole, getByLabelText } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);

  expect(getAllByRole('option').length).toBe(4);
});

test('select items should have values', () => {
  const { getAllByRole, getByLabelText } = render(SelectMock);
  const input = getByLabelText('Countries');
  fireEvent.focus(input);

  const options = getAllByRole('option');

  expect(options[0].getAttribute('value')).toBe('us');
  expect(options[1].getAttribute('value')).toBe('mx');
  expect(options[2].getAttribute('value')).toBe('ca');
  expect(options[3].getAttribute('value')).toBe('en');
});

test('select items should be unfiltered when opened', () => {
  const { getAllByRole, getByRole } = render(
    <Select onItemChange={onItemChange} label="Countries" placeholder="Choose country" value="mx">
      <Select.Option value="us">United States</Select.Option>
      <Select.Option value="mx">Mexico</Select.Option>
      <Select.Option value="ca">Canada</Select.Option>
      <Select.Option value="en">England</Select.Option>
    </Select>,
  );
  const button = getByRole('button');

  fireEvent.click(button);

  const options = getAllByRole('option');

  expect(options.length).toBe(4);
});

test('select item should be highlighted when opened', () => {
  const { getAllByRole, getByLabelText, getByRole } = render(
    <Select onItemChange={onItemChange} label="Countries" placeholder="Choose country" value="mx">
      <Select.Option value="us">United States</Select.Option>
      <Select.Option value="mx">Mexico</Select.Option>
      <Select.Option value="ca">Canada</Select.Option>
      <Select.Option value="en">England</Select.Option>
    </Select>,
  );
  const button = getByRole('button');
  const input = getByLabelText('Countries');

  fireEvent.click(button);

  const options = getAllByRole('option');

  expect(options[1].dataset.highlighted).toBe('true');
  expect(input.getAttribute('aria-activedescendant')).toEqual(options[1].id);
});

test('select items should be filterable', async () => {
  const { getAllByRole, getByLabelText, getByRole } = render(SelectMock);
  const button = getByRole('button');

  fireEvent.click(button);
  fireEvent.change(getByLabelText('Countries'), { target: { value: 'm' } });

  const options = getAllByRole('option');

  expect(options.length).toBe(1);
});

test('up/down arrows should change select item selection', () => {
  const { getAllByRole, getByLabelText } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);

  const options = getAllByRole('option');

  fireEvent.keyDown(input, { key: 'ArrowDown' });
  expect(options[0].dataset.highlighted).toBe('true');
  expect(input.getAttribute('aria-activedescendant')).toEqual(options[0].id);

  fireEvent.keyDown(input, { key: 'ArrowDown' });
  expect(options[1].dataset.highlighted).toBe('true');
  expect(input.getAttribute('aria-activedescendant')).toEqual(options[1].id);

  fireEvent.keyDown(input, { key: 'ArrowUp' });
  expect(options[0].dataset.highlighted).toBe('true');
  expect(input.getAttribute('aria-activedescendant')).toEqual(options[0].id);
});

test('esc should close menu', () => {
  const { getByLabelText, queryByRole } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);
  expect(queryByRole('listbox')).toBeInTheDocument();

  fireEvent.keyDown(input, { key: 'Escape' });
  expect(queryByRole('listbox')).not.toBeInTheDocument();
});

test('tab should close menu', () => {
  const { getByLabelText, queryByRole } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);
  expect(queryByRole('listbox')).toBeInTheDocument();

  fireEvent.keyDown(input, { key: 'Tab' });
  expect(queryByRole('listbox')).not.toBeInTheDocument();
});

test('home should select first select item', () => {
  const { getAllByRole, getByLabelText } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);

  const options = getAllByRole('option');

  fireEvent.keyDown(input, { key: 'ArrowDown' });
  fireEvent.keyDown(input, { key: 'ArrowDown' });
  fireEvent.keyDown(input, { key: 'ArrowDown' });
  expect(options[2].dataset.highlighted).toBe('true');

  fireEvent.keyDown(input, { key: 'Home' });
  expect(options[0].dataset.highlighted).toBe('true');
  expect(input.getAttribute('aria-activedescendant')).toEqual(options[0].id);
});

test('end should select last select item', () => {
  const { getAllByRole, getByLabelText } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);

  const options = getAllByRole('option');

  fireEvent.keyDown(input, { key: 'ArrowDown' });
  expect(options[0].dataset.highlighted).toBe('true');

  fireEvent.keyDown(input, { key: 'End' });
  expect(options[3].dataset.highlighted).toBe('true');
  expect(input.getAttribute('aria-activedescendant')).toEqual(options[3].id);
});

test('enter should trigger onItemChange', () => {
  const { getByLabelText } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);
  fireEvent.keyDown(input, { key: 'ArrowDown' });
  fireEvent.keyDown(input, { key: 'Enter' });
  expect(onItemChange).toHaveBeenCalledWith('us');
});

test('space should trigger onItemChange', () => {
  const { getByLabelText } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);
  fireEvent.keyDown(input, { key: 'ArrowDown' });
  fireEvent.keyDown(input, { key: 'ArrowDown' });
  fireEvent.keyDown(input, { key: ' ' });
  expect(onItemChange).toHaveBeenCalledWith('mx');
});

test('clicking on select options should trigger onItemClick', () => {
  const { getAllByRole, getByLabelText } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);

  const options = getAllByRole('option');

  fireEvent.mouseOver(options[1]);
  fireEvent.click(options[1]);
  expect(onItemChange).toHaveBeenCalledWith('mx');
});

test('select options should be highlighted when moused over', () => {
  const { getByLabelText, getByRole } = render(SelectMock);
  const input = getByLabelText('Countries');

  fireEvent.focus(input);

  const option = getByRole('option');

  fireEvent.keyDown(input, { key: 'ArrowDown' });
  fireEvent.mouseOver(option);
  expect(option.dataset.highlighted).toBe('true');
});
