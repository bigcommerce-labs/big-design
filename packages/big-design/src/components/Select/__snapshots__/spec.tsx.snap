// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`select action supports actionTypes 1`] = `
.c0 {
  border-top: 1px solid #D9DCE9;
  color: #DB3643;
  margin-top: 0.5rem;
  min-width: 16rem;
  outline: none;
  overflow: hidden;
  padding: 0.5rem 0 0;
}

.c0[data-highlighted='true'] > div {
  background-color: #F6F7FC;
}

.c0:only-child {
  border: none;
  margin: 0;
  padding: 0;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.25rem;
  padding: 0 1rem;
  width: 100%;
}

.c2 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

<li
  aria-selected="true"
  class="c0"
  data-highlighted="false"
  id="select_2-item-4"
  role="option"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    >
      Action
    </span>
  </div>
</li>
`;

exports[`select action supports icons 1`] = `
.c0 {
  border-top: 1px solid #D9DCE9;
  color: #3C64F4;
  margin-top: 0.5rem;
  min-width: 16rem;
  outline: none;
  overflow: hidden;
  padding: 0.5rem 0 0;
}

.c0[data-highlighted='true'] > div {
  background-color: #F6F7FC;
}

.c0:only-child {
  border: none;
  margin: 0;
  padding: 0;
}

.c1 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: 2.25rem;
  padding: 0 1rem;
  width: 100%;
}

.c2 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: inline-grid;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
}

<li
  aria-selected="true"
  class="c0"
  data-highlighted="false"
  id="select_2-item-4"
  role="option"
>
  <div
    class="c1"
  >
    <span
      class="c2"
    >
      <svg
        fill="none"
        height="1.25rem"
        style="width: 1.25rem; height: 1.25rem;"
        viewBox="0 0 20 20"
        width="1.25rem"
      >
        <title />
        <path
          d="M16 11h-5v5c0 .55-.45 1-1 1s-1-.45-1-1v-5H4c-.55 0-1-.45-1-1s.45-1 1-1h5V4c0-.55.45-1 1-1s1 .45 1 1v5h5c.55 0 1 .45 1 1s-.45 1-1 1z"
          fill="currentColor"
        />
      </svg>
      Action
    </span>
  </div>
</li>
`;
